image_client/picturae_csv_create.py

```diff
@@ -0,0 +1,472 @@
+                    hybrid_base = full_name
+                    full_name = " ".join(taxon_strings[:2])
+                else:
+                    hybrid_base = full_name
+                    full_name = taxon_strings[0]
+                    if full_name != genus:
+                        full_name = taxon_strings[-1]
+
+            elif len(first_intra) != len(full_name):
+                if "var." in full_name or "subsp." in full_name or " f." in full_name or "subf." in full_name:
+                    hybrid_base = full_name
+                    full_name = " ".join(taxon_strings[:4])
+                else:
+                    pass
+
+        return str(gen_spec), str(full_name), str(first_intra), str(tax_name), str(hybrid_base)
+
+
+    def taxon_unmatch_create(self, unmatched_taxa: pd.DataFrame):
+        """taxon_unmatch_create: creates sql query for creating new records in taxa unmatch,
+                                from rows that did not pass TNRS successfully,
+                                either through spelling, or taxonomic errors.
+            args:
+                unmatched_taxa: a pandas dataframe with unmatched taxa terms filtered by score
+        """
+        print("uploading unmatched taxa")
+        for index, row in unmatched_taxa.iterrows():
+            catalognumber = unmatched_taxa.columns.get_loc("CatalogNumber")
+
+            sql = create_tnrs_unmatch_tab(row=row, df=unmatched_taxa, tab_name='taxa_unmatch')
+
+            sql_result = get_one_match(connection=self.specify_db_connection, tab_name='taxa_unmatch',
+                                      id_col='CatalogNumber', key_col='CatalogNumber',
+                                      match=row[catalognumber], match_type='integer')
+            if sql_result is None:
+                insert_table_record(connection=self.specify_db_connection, logger_int=self.logger, sql=sql)
+            else:
+                pass
+
+    def taxon_check_real(self):
+        """taxon_check_real:
+           Sends the concatenated taxon column, through TNRS, to match names,
+           with and without spelling mistakes, only checks base name
+           for hybrids as IPNI does not work well with hybrids
+           """
+
+        bar_tax = self.record_full[['CatalogNumber', 'fullname']]
+
+        pandas2ri.activate()
+
+        r_dataframe_tax = pandas2ri.py2rpy(bar_tax)
+
+        robjects.globalenv['r_dataframe_taxon'] = r_dataframe_tax
+
+        with open('taxon_check/R_TNRS.R', 'r') as file:
+            r_script = file.read()
+
+        robjects.r(r_script)
+
+        resolved_taxon = robjects.r['resolved_taxa']
+
+        resolved_taxon = robjects.conversion.rpy2py(resolved_taxon)
+
+        # filtering out taxa without exact matches , saving to db table
+        unmatched_taxa = resolved_taxon[resolved_taxon["overall_score"] < .99]
+
+        # writing unmatched taxa to db table taxa_unmatch
+
+        if len(unmatched_taxa) > 0:
```

Comment:
replaced block of sql insert code with function called from sql_csv_utils, reduced length of function